### ISSUE TITLE
upspin: browser integration plan

### ISSUE THREAD

**Issue by adg** [upspin/upspin#410]
A frustration that I have with Upspin right now is that there's no easy way for me to send someone a link to something in Upspin space. This issue is a proposal for a system that makes it easy to share with anyone links to contentn in Upspin space.

---

Part one:

Develop a web UI for Upspin for a convenient and cross-platform way to work with Upspin files. This would augment `upspinfs` and the `upspin` command-line tool, not replace them. It would look something like this:

- A server daemon runs on user's workstation, serving HTTP.
- The web UI shows Upspin files with a pretty file explorer. It can:
  - Drag and drop files into the browser,
  - Efficiently copy files within Upspin space (without making copies of the references).
- When a user does a read or write to Upspin space, the server daemon makes Upspin API calls on behalf of the user (the user's keys are not known to the browser).
- The server is aware of `upspinfs`, and has a button to 'open file' or 'open directory' that uses  `file:///my/upsipnfs/root/` links to load the content using local apps (or the OS's built in file browser) through `upspinfs`.

Part two:

Run an instance of the web ui on upspin.io (frontend) that can show all public files (in any tree), but only serves file contents for specific trees (`release@upspin.io`, for example).

So, if you try to access `https://upspin.io/user@example.com/private/file` the web UI prompts you to install a Chrome extension to access the data.

Part three:

Develop a Chrome extension that hijacks requests to `https://upspin.io/*@*/*`, and for those requests it serves them using the local web UI daemon (developed in part one). This means that even though it appears that you are using a service https://upspin.io/, you're really fetching data using your local Upspin client, credentials, and cache.

The Chrome extension would probably use Chrome's [Native Messaging](https://developer.chrome.com/extensions/nativeMessaging) protocol to speak to a Go daemon that makes Upspin requests. This means that no keys are seen by the browser, while still being secure in that no browser code outside the extension can speak to the Upspin daemon.

---

The upshot of all this is that anyone can refer to a file in Upspin space using a URL with the prefix `https://upspin.io/`. If they don't have Upspin installed, then the server at upspin.io will tell them how to get set up. If they do have Upspin then they can access the files right there in the browser, or click a button to open the files with a native application.

I think this would solve a significant bootstrapping hurdle, and actually lead to a much nicer experience than what we see with similar tools like DropBox or Google Drive.

**Comment by adg** [comment]
Relates to #304.

**Comment by adg** [comment]
Status update:

We have the beginnings of an [Upspin browser](https://godoc.org/exp.upspin.io/cmd/browser/). There are more features to add, but you can list, read, copy, and delete files. It has a [Norton Commander](https://en.wikipedia.org/wiki/Norton_Commander)-style dual browser setup.

**Comment by adg** [comment]
Status update:

The [browser](https://godoc.org/exp.upspin.io/cmd/browser/) now includes a user signup and server deployment helper. I sent [a call for beta testers](https://groups.google.com/forum/#!msg/upspin/HHxZxJq3MnM/QEDOR9iIBgAJ) yesterday.

**Comment by srenatus** [comment]
⏩ I suppose its URL now is https://godoc.org/augie.upspin.io/cmd/upspin-ui

**Comment by adg** [comment]
Yes, and the upspin-ui is now included in binary downloads, and is now the encouraged setup path.

**Comment by srenatus** [comment]
@adg Oh indeed. 😄 Sorry for the noise.
